### PR TITLE
NRPT-758: updated front ends to accept arrays

### PR DESCRIPTION
**Pull request by MaxWardle**
[https://bcmines.atlassian.net/browse/NRPT-758](https://bcmines.atlassian.net/browse/NRPT-758)

Api changes in [NRPT-757](https://github.com/bcgov/NRPTI/pull/846)

### CODE REVIEW

File: angular/projects/admin-nrpti/src/app/records/administrative-sanctions/administrative-sanction-detail/administrative-sanction-detail.component.html [hunk 1]
```diff
@@ -42,19 +42,34 @@ <h4 class="grid-section-header-title-text blue-header">Shared Data [Master]</h4>
             <label class="grid-item-name">Issuing Agency</label>
             <span class="grid-item-value">{{ (data && data._master.issuingAgency) || '-' }}</span>
           </div>
-          <div class="grid-item__row">
+          <!-- <div class="grid-item__row">
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/administrative-sanctions/administrative-sanction-detail/administrative-sanction-detail.component.html:45]: remove commented code

File: angular/projects/admin-nrpti/src/app/records/certificate-amendments/certificate-amendments-detail/certificate-amendments-detail.component.html [hunk 2]
```diff
@@ -42,20 +42,34 @@ <h4 class="grid-section-header-title-text blue-header">Shared Data [Master]</h4>
             <label class="grid-item-name">Issuing Agency</label>
             <span class="grid-item-value">{{ (data && data._master.issuingAgency) || '-' }}</span>
           </div>
-          <div class="grid-item__row">
+          <!-- <div class="grid-item__row">
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/certificate-amendments/certificate-amendments-detail/certificate-amendments-detail.component.html:45]: remove

File: angular/projects/admin-nrpti/src/app/records/certificates/certificate-add-edit/certificate-add-edit.component.ts [hunk 3]
```diff
@@ -2,9 +2,9 @@ import { Component, OnInit, ChangeDetectorRef, OnDestroy } from '@angular/core';
 import { ActivatedRoute, Router } from '@angular/router';
 import { takeUntil } from 'rxjs/operators';
 import { Subject } from 'rxjs';
-import { FormGroup, FormControl } from '@angular/forms';
+import { FormGroup, FormControl, FormArray } from '@angular/forms';
 import { Picklists, EpicProjectIds } from '../../../../../../common/src/app/utils/record-constants';
-import { Legislation } from '../../../../../../common/src/app/models/master/common-models/legislation';
+// import { Legislation } from '../../../../../../common/src/app/models/master/common-models/legislation';
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/certificates/certificate-add-edit/certificate-add-edit.component.ts:7]: remove

File: angular/projects/admin-nrpti/src/app/records/correspondences/correspondence-add-edit/correspondence-add-edit.component.ts [hunk 4]
```diff
@@ -274,6 +222,81 @@ export class CorrespondenceAddEditComponent implements OnInit, OnDestroy {
...
+    });
+    return legislations;
+  }
+
+
+  /**
+   * Parses an array of penalties FormGroups into objects expected by the API.
+   *
+   * @returns {object[]} array of penalties objects
+   * @memberof CourtConvictionAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/correspondences/correspondence-add-edit/correspondence-add-edit.component.ts:276]: nit: update memberof

File: angular/projects/admin-nrpti/src/app/records/inspections/inspection-add-edit/inspection-add-edit.component.ts [hunk 5]
```diff
@@ -303,6 +252,83 @@ export class InspectionAddEditComponent implements OnInit, OnDestroy {
     });
   }
 
+  /**
+   * Builds an array of legislations FormGroups, each with its own set of FormControls.
+   *
+   * @returns {FormGroup[]} array of legislations FormGroup elements
+   * @memberof AdministrativePenaltyAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/inspections/inspection-add-edit/inspection-add-edit.component.ts:259]: nit: update memberOf

File: angular/projects/admin-nrpti/src/app/records/orders/order-add-edit/order-add-edit.component.ts [hunk 6]
```diff
@@ -310,6 +259,83 @@ export class OrderAddEditComponent implements OnInit, OnDestroy {
...
+    });
+    return legislations;
+  }
+
+
+  /**
+   * Parses an array of penalties FormGroups into objects expected by the API.
+   *
+   * @returns {object[]} array of penalties objects
+   * @memberof CourtConvictionAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/orders/order-add-edit/order-add-edit.component.ts:313]: nit: update

File: angular/projects/admin-nrpti/src/app/records/reports/report-add-edit/report-add-edit.component.ts [hunk 7]
```diff
@@ -275,6 +223,83 @@ export class ReportAddEditComponent implements OnInit, OnDestroy {
     });
   }
 
+  /**
+   * Builds an array of legislations FormGroups, each with its own set of FormControls.
+   *
+   * @returns {FormGroup[]} array of legislations FormGroup elements
+   * @memberof AdministrativePenaltyAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/reports/report-add-edit/report-add-edit.component.ts:230]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/reports/report-add-edit/report-add-edit.component.ts [hunk 8]
```diff
@@ -275,6 +223,83 @@ export class ReportAddEditComponent implements OnInit, OnDestroy {
...
+    });
+    return legislations;
+  }
+
+
+  /**
+   * Parses an array of penalties FormGroups into objects expected by the API.
+   *
+   * @returns {object[]} array of penalties objects
+   * @memberof CourtConvictionAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/reports/report-add-edit/report-add-edit.component.ts:277]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/restorative-justices/restorative-justice-add-edit/restorative-justice-add-edit.component.ts [hunk 9]
```diff
@@ -273,6 +223,83 @@ export class RestorativeJusticeAddEditComponent implements OnInit, OnDestroy {
...
+    });
+    return legislations;
+  }
+
+
+  /**
+   * Parses an array of penalties FormGroups into objects expected by the API.
+   *
+   * @returns {object[]} array of penalties objects
+   * @memberof CourtConvictionAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/restorative-justices/restorative-justice-add-edit/restorative-justice-add-edit.component.ts:277]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/restorative-justices/restorative-justice-add-edit/restorative-justice-add-edit.component.ts [hunk 10]
```diff
@@ -273,6 +223,83 @@ export class RestorativeJusticeAddEditComponent implements OnInit, OnDestroy {
     });
   }
 
+  /**
+   * Builds an array of legislations FormGroups, each with its own set of FormControls.
+   *
+   * @returns {FormGroup[]} array of legislations FormGroup elements
+   * @memberof AdministrativePenaltyAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/restorative-justices/restorative-justice-add-edit/restorative-justice-add-edit.component.ts:230]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/self-reports/self-report-add-edit/self-report-add-edit.component.ts [hunk 11]
```diff
@@ -206,6 +155,83 @@ export class SelfReportAddEditComponent implements OnInit, OnDestroy {
     });
   }
 
+  /**
+   * Builds an array of legislations FormGroups, each with its own set of FormControls.
+   *
+   * @returns {FormGroup[]} array of legislations FormGroup elements
+   * @memberof AdministrativePenaltyAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/self-reports/self-report-add-edit/self-report-add-edit.component.ts:162]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/self-reports/self-report-add-edit/self-report-add-edit.component.ts [hunk 12]
```diff
@@ -206,6 +155,83 @@ export class SelfReportAddEditComponent implements OnInit, OnDestroy {
...
+    });
+    return legislations;
+  }
+
+
+  /**
+   * Parses an array of penalties FormGroups into objects expected by the API.
+   *
+   * @returns {object[]} array of penalties objects
+   * @memberof CourtConvictionAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/self-reports/self-report-add-edit/self-report-add-edit.component.ts:209]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/tickets/ticket-add-edit/ticket-add-edit.component.ts [hunk 13]
```diff
@@ -273,6 +223,82 @@ export class TicketAddEditComponent implements OnInit, OnDestroy {
     });
   }
 
+  /**
+   * Builds an array of legislations FormGroups, each with its own set of FormControls.
+   *
+   * @returns {FormGroup[]} array of legislations FormGroup elements
+   * @memberof AdministrativePenaltyAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/tickets/ticket-add-edit/ticket-add-edit.component.ts:230]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/tickets/ticket-add-edit/ticket-add-edit.component.ts [hunk 14]
```diff
@@ -273,6 +223,82 @@ export class TicketAddEditComponent implements OnInit, OnDestroy {
...
+    });
+    return legislations;
+  }
+
+
+  /**
+   * Parses an array of penalties FormGroups into objects expected by the API.
+   *
+   * @returns {object[]} array of penalties objects
+   * @memberof CourtConvictionAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/tickets/ticket-add-edit/ticket-add-edit.component.ts:277]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/warnings/warning-add-edit/warning-add-edit.component.ts [hunk 15]
```diff
@@ -281,6 +230,82 @@ export class WarningAddEditComponent implements OnInit, OnDestroy {
...
+    });
+    return legislations;
+  }
+
+
+  /**
+   * Parses an array of penalties FormGroups into objects expected by the API.
+   *
+   * @returns {object[]} array of penalties objects
+   * @memberof CourtConvictionAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/warnings/warning-add-edit/warning-add-edit.component.ts:284]: nit: update name

File: angular/projects/admin-nrpti/src/app/records/warnings/warning-add-edit/warning-add-edit.component.ts [hunk 16]
```diff
@@ -281,6 +230,82 @@ export class WarningAddEditComponent implements OnInit, OnDestroy {
     });
   }
 
+  /**
+   * Builds an array of legislations FormGroups, each with its own set of FormControls.
+   *
+   * @returns {FormGroup[]} array of legislations FormGroup elements
+   * @memberof AdministrativePenaltyAddEditComponent
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/records/warnings/warning-add-edit/warning-add-edit.component.ts:237]: nit: update name

**Review by mtCarto — APPROVED** [review]
Code looks great, comments are just some tidying and cleanup.

**Review by BcGovNeal — CHANGES_REQUESTED** [review]
There is a bug with Add Legislation button.  

When editing records where `sourceSystemRef !== 'nrpti'`, the form controls are disabled.  Then when you go add a new legislation, the FormArray deletes FormGroups that are disabled.   This causes the original legislation to be removed when you add a new legislation.

The disabled values are still available through `getRawValue()` instead of `value`.  But I think in this case we should just disable the Add button if the legislations are not supposed to be edited.

**Comment by MaxWardle** [comment]
> There is a bug with Add Legislation button.
> 
> When editing records where `sourceSystemRef !== 'nrpti'`, the form controls are disabled. Then when you go add a new legislation, the FormArray deletes FormGroups that are disabled. This causes the original legislation to be removed when you add a new legislation.
> 
> The disabled values are still available through `getRawValue()` instead of `value`. But I think in this case we should just disable the Add button if the legislations are not supposed to be edited.

Good catch, thanks!

File: angular/projects/admin-nrpti/src/app/records/warnings/warning-add-edit/warning-add-edit.component.ts (hunk 16 above)
Comment (MaxWardle) [reply angular/projects/admin-nrpti/src/app/records/warnings/warning-add-edit/warning-add-edit.component.ts]: Missed these, thank you!

**Review by BcGovNeal — APPROVED** [review]

**Pull request merged**